File: app/src/main/java/id/my/berviantoleo/ecceg_rsa_app/fragment/RSAGenerateKeyFragment.java
package id.my.berviantoleo.ecceg_rsa_app.fragment;

import android.Manifest;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import id.my.berviantoleo.ecceg_rsa_app.R;
import id.my.berviantoleo.ecceg_rsa_app.lib.rsa.RSA;
import com.google.android.material.textfield.TextInputEditText;
import com.gun0912.tedpermission.PermissionListener;
import com.gun0912.tedpermission.TedPermission;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.List;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cc.cloudist.acplibrary.ACProgressFlower;


public class RSAGenerateKeyFragment extends Fragment {

    @BindView(R.id.public_location_save)
    protected TextInputEditText publicLocation;
    @BindView(R.id.private_location_save)
    protected TextInputEditText privateLocation;
    @BindView(R.id.byte_size)
    protected TextInputEditText byteSize;
    private ACProgressFlower loadingView;
    private PermissionListener extract;

    public RSAGenerateKeyFragment() {
        // Required empty public constructor
    }

    public static RSAGenerateKeyFragment newInstance() {
        return new RSAGenerateKeyFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_rsagenerate_key, container, false);
        ButterKnife.bind(this, view);
        loadingView = new ACProgressFlower.Builder(getContext()).build();
        loadingView.setCanceledOnTouchOutside(false);
        loadingView.setCancelable(false);
        extract = new PermissionListener() {
            @Override
            public void onPermissionGranted() {
                File location = Environment.getExternalStorageDirectory();
                File newLocation = new File(location, "RSA/");
                if (!newLocation.exists()) {
                    newLocation.mkdir();
                }
                loadingView.show();
                new GenerateKey(RSAGenerateKeyFragment.this).execute(Objects.requireNonNull(byteSize.getText()).toString(), newLocation.getAbsolutePath() + "/" + privateLocation.getText().toString(),
                        newLocation.getAbsolutePath() + "/" + Objects.requireNonNull(publicLocation.getText()).toString());
            }

            @Override
            public void onPermissionDenied(List<String> deniedPermissions) {
                Toast.makeText(getContext(), "Permission Denied\n" + deniedPermissions.toString(), Toast.LENGTH_SHORT).show();
            }
        };
        return view;
    }

    @OnClick(R.id.generate_button)
    void generateKey() {
        if (!Objects.requireNonNull(byteSize.getText()).toString().equalsIgnoreCase("")
                && !Objects.requireNonNull(privateLocation.getText()).toString().equalsIgnoreCase("")
                && !Objects.requireNonNull(publicLocation.getText()).toString().equalsIgnoreCase("")) {
            if (Integer.valueOf(byteSize.getText().toString()) >= 1024) {
                TedPermission.with(Objects.requireNonNull(getContext()))
                        .setPermissionListener(extract)
                        .setDeniedMessage("If you reject permission,you can not use this service\n\nPlease turn on permissions at [Setting] > [Permission]")
                        .setPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                        .check();
            }
        }
    }

    private class GenerateKey extends AsyncTask<String, Integer, Void> {

        private WeakReference<RSAGenerateKeyFragment> activityReference;

        // only retain a weak reference to the activity
        GenerateKey(RSAGenerateKeyFragment context) {
            activityReference = new WeakReference<>(context);
        }

        @Override
        protected Void doInBackground(String... strings) {
            int byteSize = Integer.valueOf(strings[0]);
            RSA.generateKey(byteSize, strings[1], strings[2]);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {

            RSAGenerateKeyFragment activity = activityReference.get();
            if (activity == null) return;
            activity.loadingView.dismiss();
            Toast.makeText(getContext(), "Finished", Toast.LENGTH_SHORT).show();
        }
    }

}

File: app/src/main/java/id/my/berviantoleo/ecceg_rsa_app/fragment/RSAGenerateKeyFragment.kt
package id.my.berviantoleo.ecceg_rsa_app.fragment

import android.Manifest
import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.gun0912.tedpermission.PermissionListener
import com.gun0912.tedpermission.normal.TedPermission
import id.my.berviantoleo.ecceg_rsa_app.R
import id.my.berviantoleo.ecceg_rsa_app.databinding.FragmentRsagenerateKeyBinding
import id.my.berviantoleo.ecceg_rsa_app.lib.rsa.RSA
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class RSAGenerateKeyFragment : Fragment() {

    private var _binding: FragmentRsagenerateKeyBinding? = null
    private val binding get() = _binding!!

    private var loadingDialog: AlertDialog? = null
    private var permissionListener: PermissionListener? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentRsagenerateKeyBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupLoadingDialog()
        setupPermissionListener()

        binding.generateButton.setOnClickListener {
            validateAndGenerateKeys()
        }
    }

    private fun setupLoadingDialog() {
        val builder = AlertDialog.Builder(requireContext())
        builder.setCancelable(false)
        builder.setView(R.layout.layout_loading_dialog)
        loadingDialog = builder.create()
    }

    private fun setupPermissionListener() {
        permissionListener = object : PermissionListener {
            override fun onPermissionGranted() {
                val byteSizeStr = binding.byteSize.text.toString()
                val privateKeyFileName = binding.privateLocationSave.text.toString()
                val publicKeyFileName = binding.publicLocationSave.text.toString()

                val byteSizeInt = byteSizeStr.toIntOrNull()

                if (byteSizeInt == null) {
                    Toast.makeText(context, "Invalid byte size entered.", Toast.LENGTH_SHORT).show()
                    return
                }

                lifecycleScope.launch {
                    generateAndSaveKeys(byteSizeInt, privateKeyFileName, publicKeyFileName)
                }
            }

            override fun onPermissionDenied(deniedPermissions: List<String>) {
                Toast.makeText(context, "Permission Denied\n$deniedPermissions", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun validateAndGenerateKeys() {
        val byteSizeStr = binding.byteSize.text.toString()
        val privateKeyFileName = binding.privateLocationSave.text.toString()
        val publicKeyFileName = binding.publicLocationSave.text.toString()

        if (publicKeyFileName.isBlank() || privateKeyFileName.isBlank() || byteSizeStr.isBlank()) {
            Toast.makeText(context, "All fields must be filled.", Toast.LENGTH_SHORT).show()
            return
        }

        val byteSizeInt = byteSizeStr.toIntOrNull()
        if (byteSizeInt == null || byteSizeInt < 1024) {
            Toast.makeText(context, "Key size must be a number and at least 1024.", Toast.LENGTH_SHORT).show()
            return
        }

        TedPermission.create()
            .setPermissionListener(permissionListener)
            .setDeniedMessage("If you reject permission, you cannot use this service\n\nPlease turn on permissions at [Setting] > [Permission]")
            .setPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE)
            .check()
    }

    private suspend fun generateAndSaveKeys(byteSize: Int, privateKeyFileName: String, publicKeyFileName: String) {
        withContext(Dispatchers.Main) {
            loadingDialog?.show()
        }

        try {
            val location = Environment.getExternalStorageDirectory()
            val rsaDir = File(location, "RSA")
            if (!rsaDir.exists()) {
                rsaDir.mkdirs()
            }

            val privateKeyPath = File(rsaDir, privateKeyFileName).absolutePath
            val publicKeyPath = File(rsaDir, publicKeyFileName).absolutePath

            withContext(Dispatchers.IO) {
                RSA.generateKey(byteSize, privateKeyPath, publicKeyPath)
            }

            withContext(Dispatchers.Main) {
                loadingDialog?.dismiss()
                Toast.makeText(context, "Keys generated successfully!", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                loadingDialog?.dismiss()
                Toast.makeText(context, "Error generating keys: ${e.message}", Toast.LENGTH_LONG).show()
            }
            e.printStackTrace()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        loadingDialog?.dismiss() // Ensure dialog is dismissed to prevent leaks
        _binding = null
    }

    companion object {
        @JvmStatic
        fun newInstance(): RSAGenerateKeyFragment {
            return RSAGenerateKeyFragment()
        }
    }
}
